caddyhttp/proxy/reverseproxy.go

```diff
@@ -154,7 +155,12 @@ func NewSingleHostReverseProxy(target *url.URL, without string, keepalive int) *
 				prefer(target.RawPath, target.Path),
 				prefer(req.URL.RawPath, req.URL.Path))
 		}
-		req.URL.Path = singleJoiningSlash(target.Path, req.URL.Path)
+		req.URL.Path = singleJoiningSlash(target.Path,
+			prefer(req.Context().Value(staticfiles.URLPathCtxKey).(string), req.URL.Path))
+		// req.URL.Path must be consistent with decoded form of req.URL.RawPath if any
+		if req.URL.RawPath != "" && req.URL.RawPath != req.URL.EscapedPath() {
+			panic("RawPath doesn't match Path")
```

Comment:
@Cinerar Because we have sanitize the original request's URL ([here](https://github.com/mholt/caddy/blob/master/caddyhttp/httpserver/server.go#L297). So if the original request's URL is `/%2F/test`, it will be changed to `/test` before proxy middleware. This is the reason why some tests don't catch.